proposals/ParameterUpdate.md

```diff
@@ -0,0 +1,687 @@
+
+# Training data (inputs/outputs)
+x = tf.placeholder(dtype=tf.float32)
+y = tf.placeholder(dtype=tf.float32)
+x_train = [1, 2, 3, 4]
+y_train = [0, 1, 2, 3]
+
+linear_model = W * x + b
+squared_deltas = tf.square(linear_model - y_train)
+loss = tf.reduce_sum(squared_deltas)
+
+# Optimizer
+optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.01)
+train = optimizer.minimize(loss=loss)
+```
+
+In the last line above: how does the optimizer determine which tensors are parameters to be minimized? This is done implicitly by examining the graph of `loss`: since the only Variables in the graph are `W` and `B`, they are determined to be the parameters and are minimized.
+
+In Swift, TensorFlow graphs are an implementation detail and aren't visible to users: there's no way to inspect whether tensors are placeholders/constants/variables, so the TensorFlow style of implicit parameter analysis is not really suitable. With implicit parameters, it's difficult to work with parameters directly (e.g. to implement a custom optimizer for arbitrary parameters). The authors believe that parameter representation and aggregate parameter update are language-design problems and should be explicitly clear in Swift.
+
+## Parameter update in Swift
+
+The current Swift parameter update design is based on two protocols: `ParameterAggregate` and `Parameterized`.
+
+They enable:
+* Parameter update for parameter aggregates and parameterized models, where all parameters have the same type.
+* Nested parameter aggregates and parameterized models as parameters, which is important for layer-based high level APIs.
+* General optimizers that work with any floating-point parameter aggregate type.
+
+Examples:
+* [MNIST](https://github.com/tensorflow/swift-models/blob/master/MNIST/MNIST.swift): demonstrates `ParameterAggregate`.
+* [Autoencoder](https://github.com/tensorflow/swift-models/blob/master/Autoencoder/Autoencoder.swift): demonstrates `Parameterized`.
+* Optimizers are a work-in-progress, see a sample sketch in this [PR description](https://github.com/apple/swift/pull/18171#issue-203387822).
+
+Implementation:
+* [Part 1: ParameterAggregate protocol and synthesis](https://github.com/apple/swift/pull/18140)
+* [Part 2: Parameterized protocol and synthesis](https://github.com/apple/swift/pull/18171)
+
+---
+
+Two protocols have been added to the TensorFlow module: `ParameterAggregate` and `Parameterized`.
+
+The `ParameterAggregate` protocol represents an aggregate of parameters. Types that conform to `ParameterAggregate` must specify a `Parameter` associated type and an update method.
+
+```swift
+/// A type representing an aggregate of parameters.
+public protocol ParameterAggregate {
+  /// The parameter type.
+  associatedtype Parameter
+
+  /// Update parameters with their gradient values, using an updater function.
+  mutating func update(
+    withGradients gradients: Self,
+    _ updater: (inout Parameter, Parameter) -> Void
```

Comment:
Random API naming thing: the name for closure arguments is usually a verb or a verb phrase. It makes it clear when you call this closure in the implementation as well. Examples:
```swift
extension Sequence {
  func filter(_ isIncluded: ...) -> ...
  func map(_ transform: ...) -> ...
}
```

Addressing this is not important and doesn't have to be in the doc.

Reply:
Thanks for bringing this up! Could you suggest a more appropriate name?

Reply:
No name is fully suitable because function types can't carry argument labels. Maybe just `update`?

Reply:
I like `update`:
```swift
update(&param, grad)
```

Reply:
Done in https://github.com/tensorflow/swift/pull/53/commits/08a21ceb187c45847b825e8aa5493500264775b3.

Is it too weird that `update` is an argument name of the `update` function (and both are functions)?
```swift
mutating func update(
  withGradients gradients: Self,
  _ update: (inout Parameter, Parameter) -> Void
)
```

Reply:
Indeed. `updateParameter`?

Reply:
Hmm, I feel it's not obvious why `updateParameter` takes two arguments solely based on the name. By comparison, `isIncluded` and `transform` make more sense as names per se.

I will go ahead with `updateParameter` though, since it better satisfies the naming conventions.

Reply:
`updateParameterBasedOnGradients` is obvious, but it's too long.

Reply:
Done in https://github.com/tensorflow/swift/pull/53/commits/cafa09487c7b5fbb193e5139b93d0e4d162f692d.